![AIMS Logo](https://www.uni-bielefeld.de/fakultaeten/mathematik/ag/hofmanova/daad-aims/aims-senegal.png)

****

# <center> <b> <span style="color:orange;"> Python Proficiency for Scientific Computing and Data Science (PyPro-SCiDaS)  </span> </b></center>

### <center> <b> <span style="color:green;">An Initiation to Programming using Python (Init2Py) </span> </b></center>
    

****
# <center> <b> <span style="color:blue;">Assignment </span> </b></center>

****
### Instructions:
You should solve these problems using only:
- Standard Python data types (e.g., str, list, dict, int).
- Built-in and user-defined functions.
- Basic flow control (if/else, for loops, while loops).
- Standard library modules such as math
****

In [3]:
import time

# Just not to repeat my self
def input_natural_number():
    is_not_int = True
    while is_not_int:
        try:
            n = int(input("Give me any natural number and I will give you the result: "))
            if n < 0:
                raise ValueError("I don't take negative number")
        except Exception as e:
            time.sleep(1)
            print(f"Error: {str(e)}")
            print("Please, I only accept natural numbers.")
            time.sleep(1)
            continue
        else:
            is_not_int = False
    
    return n

## Section 1: Python Fundamentals

### Exercise 1
Compute the following expressions for a natural number $n$ chosen by the user:

a) $$\sum_{\substack{i=0 \\ i \text{ is odd}}}^{n-1} (i + 3)$$


In [4]:
# Your code here
n = input_natural_number()
result = 0
for i in range(n):
    if i % 2 != 0:
        result += i + 3

time.sleep(1)
print(f"Here is your answer: {result}")

Give me any natural number and I will give you the result:  3


Here is your answer: 4


b) $$\prod_{\substack{i=0 \\ i \text{ is even}}}^{n-1} (i + 3)$$

In [5]:
# Your code here
n = input_natural_number()

result = 1
for i in range(n):
    if i % 2 == 0:
        result *= i + 3

time.sleep(1)
print(f"Here is your answer: {result}")

Give me any natural number and I will give you the result:  5


Here is your answer: 105


### Exercise 2
a) Write python code that computes the sum of the square of the first $n$ natural numbers. One version with a "for loop" and another version with a "while loop".

In [6]:
# Part a: For loop version
n = input_natural_number()

result = 0
for i in range(n):
    result += i ** 2

time.sleep(1)
print(f"Here is your answer: {result}")

Give me any natural number and I will give you the result:  4


Here is your answer: 14


In [7]:
# Part a: While loop version
n = input_natural_number()

result = 0
while n > 0:
    result += (n - 1) ** 2
    n = n - 1

time.sleep(1)
print(f"Here is your answer: {result}")

Give me any natural number and I will give you the result:  5


Here is your answer: 30


b) Write python code that prints, for a natural number $M$, the smallest natural number $n$ such that $1^2 + 2^2 + \cdots + n^2 \geq M$. One version with a "for loop" and another version with a "while loop".

In [8]:
# Part b: While loop version
m = input_natural_number()

sum_squares = 0
for i in range(m):
    sum_squares += i ** 2
    if sum_squares >= m:
        break

time.sleep(1)
print(f"Here is your answer: {i}") 

Give me any natural number and I will give you the result:  6


Here is your answer: 3


In [9]:
# Part b: For loop version
m = input_natural_number()

sum_squares = 0
n = 0
while sum_squares < m:
    n += 1
    sum_squares += n ** 2

time.sleep(1)
print(f"Here is your answer: {i}") 

Give me any natural number and I will give you the result:  4


Here is your answer: 3


### Exercise 3
Write a python script that asks the user for the values of $x$ and $y$ and uses these inputs to compute the numerical value of the expression:

$$\frac{\sqrt{x^2+y}}{\frac{1}{x} + 2\sin y}$$

In [15]:
import math

In [12]:
# Your code here
user_input = input("Give me two values x and y (separated by one space):")
x, y = (float(i) for i in user_input.split(' '))

result = math.sqrt(x ** 2 + y) / ((1 / x) + 2 * math.sin(y))

time.sleep(1)
print(f"Here is your answer: {result}")

Give me two values x and y (separated by one space): 4 6


Here is your answer: -15.187645717339198


### Exercise 4

A store charges XOF $12$ per item if fewer than $10$ items are purchased. If the number of items is between $10$ and $99$ (inclusive), the price is XOF $10$ per item. For $100$ items or more, the price drops to XOF $7$ per item.

a) Write a function called `Cost` that takes as input the number of items $n$ and returns the total cost.


In [36]:
# Your code here
def cost(n):
    if n < 10:
        total_cost = n * 12
    elif n >= 10 and n <= 99:
        total_cost = n * 10
    else:
        total_cost = n * 7
    return total_cost

b) Test the function `Cost` for $n = 50 \text{ and } 255$.

In [38]:
# Your code here
print(cost(50))
print(cost(255))

500
1785


### Exercise 5
For a fixed real number $x$ and a natural number $n$, we can define recursively $x^n$ using the relations:

$$x^0 = 1 \quad \text{and} \quad x^{n+1} = x \cdot x^n$$

Write a function `power(x,n)` that implements the above recursion.

In [40]:
# Your code here
def power(x, n):
    if n == 0:
        return 1
    else:
        return x * power(x, n-1)

### Exercise 6

a) Write two functions: `arithmetic_mean`, which takes two numbers and returns their [**arithmetic mean**](https://en.wikipedia.org/wiki/Arithmetic_mean)
, and `geometric_mean`, which takes two numbers and returns their [**geometric mean**](https://en.wikipedia.org/wiki/Geometric_mean). Then write a function `avggeom` that takes two input numbers and, for 10 iterations, repeatedly computes both their arithmetic mean and geometric mean, using the results from the previous iteration as the new inputs.


In [43]:
# Your code here
def arithmetic_mean(x, y):
    return (x + y) / 2

def geometric_mean(x, y):
    return (x * y) ** 0.5

def avggeom(x,y):
    for _ in range(10):
        x = arithmetic_mean(x, y)
        y = geometric_mean(x, y)
    return x, y

In [44]:
avggeom(12, 13)

(12.664924570921587, 12.664925512176977)

b) Write a function called `distancefn` that computes the [**Euclidean distance**](https://en.wikipedia.org/wiki/Euclidean_distance) between two points $(x_1, y_1)$ and $(x_2, y_2)$ in the plane. Use it in another program that computes the perimeter and the area of a triangle, given three points in the plane (each given as coordinates).

 


In [14]:
# Your code here
def distancefn(_tuple1, _tuple2):
    dist = math.sqrt((_tuple1[0] - _tuple2[0]) ** 2 + (_tuple1[1] - _tuple2[1]) ** 2)
    return dist

In [16]:
def perimeter_area(_tuple1, _tuple2, _tuple3):
    a = distancefn(_tuple1, _tuple2)
    b = distancefn(_tuple1, _tuple3)
    c = distancefn(_tuple2, _tuple3)
    p = a + b + c
    half_p = p / 2
    area = math.sqrt(half_p * (half_p - a) * (half_p - c) * (half_p - c))
    return p, area

In [17]:
perimeter_area((2, 1), (4, 1), (4, 3))

(6.82842712474619, 3.1075479480600743)

### Exercise 7

a) Let `L=[2,3,5,7,11,13,17,19,23,29,31,37,41,43,47]`. Use list comprehension to produce a list of the gaps (differences) between consecutive entries in `L`. Then find the maximum gap size and the percentage of gaps that have size $2$.

In [20]:
# Your code here
L=[2,3,5,7,11,13,17,19,23,29,31,37,41,43,47]
gaps = [L[i] - L[i-1] for i in range(1, len(L))]

# Max
print(max(gaps))

# Percentage
print(gaps.count(2) / len(gaps))

6
0.42857142857142855


b) i) Using list comprehensions, write a function `ListComp` that takes a list of strings as input and returns three objects:

- a list that consists of the strings with their first characters removed
- a list of the lengths of the strings
- a list that consists of only those strings that are at least 5 characters long


In [22]:
# Your code here
def ListComp(_list):
    removed_first_chr = [i[1:] for i in _list]
    lengths = [len(i) for i in _list]
    five_least = [i for i in _list if len(i) >= 5]

    return removed_first_chr, lengths, five_least


ii) Test your function with the list: `["hello", "world", "python", "programming", "AI", "code"]`

In [23]:
# Your code here
ListComp(["hello", "world", "python", "programming", "AI", "code"])

(['ello', 'orld', 'ython', 'rogramming', 'I', 'ode'],
 [5, 5, 6, 11, 2, 4],
 ['hello', 'world', 'python', 'programming'])

### Exercise 8

a) Write a program that repeatedly asks the user to enter product names and prices. Store all of these in a dictionary whose keys are the product names and whose values are the prices. When the user is done entering products and prices, allow them to repeatedly enter a product name and print the corresponding price or a message if the product is not in the dictionary.

In [16]:
# Your code here
products = {}
condition = True

while condition:
    product_price = input("Register the product and its price (separate them by one space): ")
    product, price = product_price.split(' ')
    products[product] = float(price)
    all_stored = input("Are you done ? (yes/no): ")
    if all_stored == 'yes':
        condition = False   

print(products)

Register the product and its price (separate them by one space):  orange 60
Are you done ? (yes/no):  no
Register the product and its price (separate them by one space):  banana 200
Are you done ? (yes/no):  yes


{'orange': 60.0, 'banana': 200.0}


In [29]:
condition = True
while condition:
    product = input("Give me the product name, I will give you its price: ")
    try:
        print(products[product])
    except KeyError:
        print(f"Oups, this product {product} is not in our store")
    last_request = input("Are you done ? (yes/no): ")
    if last_request == 'yes':
        condition = False    

Give me the product name, I will give you its price:  banana


8.0


Are you done ? (yes/no):  yes


b) Using the dictionary created above, allow the user to enter an amount and print out a list of the products whose price is less than that amount.

In [32]:
# Your code here
condition = True
filter_products = []

while condition:
    price = input("Give me a price and I'll print out a list of the products whose price is less than that amount: ")
    try:
        filter_products = [i for i, v in products.items() if v < float(price)]
    except Exception as e:
        print(f"Oooh, error: {e}")
        continue
    print(filter_products)
    condition = False

Give me a price and I'll print out a list of the products whose price is less than that amount:  10


['banana']


c) Using the dictionary created above, evaluate the most expensive product.

In [35]:
# Your code here
sorted(products, key=lambda x: x[1])[-1]

'apple'

d) Using the dictionary created above, calculate the average price of all products.

In [36]:
# Your code here
sum(products.values())/len(products)

9.0

e) Using the dictionary created above, update the prices of multiple products at once.

In [37]:
# Your code here
products.update({'apple': 20, 'banana': 15})

### Exercise 9

Use string methods and attributes to solve the following:

a) Write a function called `word_counter` that reads in the file `example.txt` and estimates the total number of words in the file. Assume words are separated by spaces.

In [39]:
# Your code here
def word_counter():
    with open('example.txt', 'r') as f:
        words_count = 0
        for line in f:
            words_count += len(words.strip().split())
    return words_count

b) Write a function called `UpLow` that reads the content of `example.txt` and then returns a dictionary containing the total number of uppercase letters and the total number of lowercase letters.

In [3]:
# Your code here
def UpLow():
    with open('example.txt', 'r') as f:
        letters = []
        for line in f:
            letters.extend(line.strip())
    ups = 0
    lows = 0
    for letter in letters:
        if letter.strip():
            if letter.isupper():
                ups += 1
            elif letter.islower():
                lows += 1
    return {'ups': ups, 'lows': lows}

c) Write a function called `count_word` that reads in `example.txt` and returns how many times the `target_word` ("the") appears in the text. Take note that the count must be case-insensitive (e.g., "The", "the", "THE" all count as one) and that it must only count whole word matches, not partial matches (e.g., "them" should not count if the target is "the").

In [52]:
# Your code here
def count_word(_pattern):
    with open('example.txt', 'r') as f:
        words_count = 0
        for line in f:
            words_count += words.strip().split().count('the')
            
    return words_count

d) Write a function called `swap_and_clean` that reads the content of the file `example.txt`. The function should process the text by converting all uppercase letters to lowercase and all lowercase letters to uppercase. It should also remove all characters that are not letters, such as numbers and punctuation, while keeping spaces intact. After performing these modifications, the function should write the resulting text to a new file called `output1.txt`.

In [7]:
# Your code here
def swap_and_clean():
    with open('example.txt', 'r') as f:
        content = f.read()

    new_content = ''
    for letter in content:
        if letter.isupper():
            new_content += letter.lower()
        if letter.islower():
            new_content += letter.upper()
    
    with open('output1.txt', 'w') as f:
        f.write(content)

e) Write a function called `palindromes` that reads the content of `example.txt`. The function should first split the text into individual words. For each word, it should be cleaned by converting it to lowercase and removing any leading or trailing punctuation. After cleaning, the function should check whether the word is a [**palindrome**](https://en.wikipedia.org/wiki/Palindrome). A palindrome is a word that reads the same forwards and backwards eg "madam". If a word is a palindrome, the function should print the original, uncleaned version of that word to the console.

In [ ]:
import re

In [8]:
# Your code here
def palindromes():
    with open('example.txt', 'r') as f:
        content = f.read()
        
    words_list = content.split(' ')
    words_list_clean = {re.sub('\W','', i.lower()): i for i in words_list}
    for k, v in words_list_clean.items():
        if k == k[::-1]:
            print(v)

In [62]:
# test
import re
re.sub('\W','', "Qwer t;")

'Qwert'

In [57]:
?re.sub

f) Write a function called `extract_and_format_numbers` that reads the file `example.txt` line by line. The function should identify any numbers that appear in the text, such as `100000000`. For each number found, it should format the number by inserting commas in the appropriate places (for example, converting `100000000` to `100,000,000`). Each formatted number should then be written on a separate line in a new file called `output2.txt`.

In [66]:
# Your code here
def extract_and_format_numbers():
    with open('example.txt', 'r') as f:
        content = f.read()
    words_list = content.split(' ')
    words_numbers_list = [i for i in words_list if i.isdigit()]
    
    num_to_save = []
    for num in words_numbers_list:
        if len(num) <= 3:
            num_to_save.append(num)
        else:
            new_num = ''
            for i, digit in enumerate(num[::-1], start=1):
                if new_num and len(new_num) % 3 == 0:
                    new_num += ','
                else:
                    new_num += digit
            new_num = new_num[::-1]
            num_to_save.append(new_num)

    with open('output2.txt', 'a') as f:
        for num in num_to_save:
            f.write(num)

In [65]:
'100'.isdigit()

True

### Exercise 10

The [**digital root**](https://en.wikipedia.org/wiki/Digital_root) of a positive number $n$ is obtained as follows: 
- Add up the digits $n$ to get a new number.
- Add up the digits of that to get another new number. Keep doing this until you get a number that has only one digit. That number is the digital root.

For example, if $n = 45893$, we add up the digits to get $$4 + 5 + 8 + 9 + 3 = 29.$$ We then add up
the digits of $29$ to get $$2 + 9 = 11.$$ We then add up the digits of $11$ to get $$1 + 1 = 2.$$ Since 2 has
only one digit, 2 is our digital root.

Write the Python code of the function `DigRoot` that returns the digital root of an integer $n$ given as input.



In [20]:
# Your code here
def DigRoot(number):
    string_number = str(number)
    if len(string_number) == 1:
        return number
    digit = 0
    while len(string_number) > 1:
        for i in string_number:
            digit += int(i)
        string_number = str(digit)
        
    return digit

DigRoot(45)

9

## Section 2: Mathematical Applications

### Exercise 1
Compute the first 30 terms of the sequence defined by $2x_{n+2} - x_{n+1} - 6x_n = 0$ sequence, given any two initial values $x_0$ and $x_1$.

In [22]:
# Your code here
def sequence(x0, x1, n=30):
    seq = [x0, x1]
    for i in range(2, n):
        x = (1/2) * (x1 + 6 * x0)
        x0, x1 = x1, x
        seq.append(x)
    return seq

sequence(0, 1)

[0,
 1,
 0.5,
 3.25,
 3.125,
 11.3125,
 15.03125,
 41.453125,
 65.8203125,
 157.26953125,
 276.095703125,
 609.8564453125,
 1133.21533203125,
 2396.177001953125,
 4597.7344970703125,
 9487.398254394531,
 18536.902618408203,
 37730.646072387695,
 74476.03089141846,
 150429.9536628723,
 298643.0695056915,
 600611.3957414627,
 1196234.906387806,
 2399951.640418291,
 4788680.539372563,
 9594195.190941155,
 19163139.213588268,
 38364155.1796176,
 76671495.2305736,
 153428213.15413958]

### Exercise 2

Let $f$ be a continuous function defined on the closed interval $[a, b]$ such that $f(a)$ and $f(b)$ have opposite signs. By the [**Intermediate Value Theorem**](https://en.wikipedia.org/wiki/Intermediate_value_theorem), there exists a root $x_0 \in (a, b)$ for which $f(x_0) = 0$.

The [**bisection method**](https://fiveable.me/numerical-analysis-i/unit-3/bisection-method-algorithm/study-guide/5LfBknp8yfSe5lse) is an iterative method for approximating the root of such a function. At each step, it divides the interval in half and selects the subinterval where the sign change occurs, repeating this process until the interval width is smaller than a given tolerance ($\epsilon$).

a) Write a function named `bisection(f, a, b, epsilon=1e-2)` that:

- Returns a tuple $(\bar{x}, f(\bar{x}))$, where $\bar{x}$ is an approximation of the root such that
$$
|\bar{x} - x_0| \le \epsilon
$$
if $f(a)\cdot f(b) < 0$.  

- Prints ``Bisection method failed'' if $f(a)\cdot f(b) \ge 0$.


In [27]:
# Your code here
def bisection(f, a, b, epsilon=1e-2):
    if f(a) * f(b) >= 0:
        print("Bisection failed")
        return
    condition = True
    while condition:
        c = (a+b)/2
        if f(a) * f(c) < 0:
            a, b = a, c
        elif f(b) * f(c) < 0:
            a, b = b, c
        if abs(a - b) <= epsilon:
            condition = False

    return b, f(b)


b) Test your function `bisection` for $ f(x) = x^3 -3x - 5$, $a=-10$ and $b=10$

In [32]:
# Your code here
f = lambda x: x ** 3 - 3 * x - 5

bisection(f, -10, 10)

(2.275390625, -0.04555877298116684)

### Exercise 3
An irrational decimal fraction, [**Champernowne constant**](https://en.wikipedia.org/wiki/Champernowne_constant), is created by concatenating the positive integers: 0.12345678910<font color='red'>1</font>112131415161718192021...

It can be seen that the 12th digit of the fractional part is <font color='red'>1</font>. If $d_n$ represents the nth digit of the fractional part, find the value of the following expression.

$$d_1 × d_{10} × d_{100} × d_{1000} × d_{10000} × d_{100000} × d_{1000000}$$

In [38]:
# Your code here
n = ''
i = 1
while len(n) < 1000000:
    n += str(i)
    i += 1
result = int(n[0]) * int(n[9]) * int(n[99]) * int(n[999]) * int(n[9999]) * int(n[99999]) * int(n[999999])
result

210

## Section 3: Advanced Programming

### Exercise 1

a) Write a Python function named `factorial2` that takes a non-negative integer $n$ as input and returns $n!$ (the factorial of $n$).

In [34]:
# Your code here
def factorial2(n):
    if n <= 1:
         return 1
    else:
        return n * factorial2(n-1)

In [35]:
factorial2(3)

6

b) i) Write a Python function named `QuadEqu` that solves a quadratic equation of the form:
$$ax^2 + bx + c = 0$$

Your function should:
- Take three real numbers $a$, $b$, and $c$ as inputs, with $a$ defaulting to 1.
- Return the real solutions of the equation (if any) in a suitable format.

In [46]:
# Your code here
import math

def QuadEqu(a, b, c):
    dis = b ** 2 - 4 * a * c
    if dis > 0:
        x1 = (-b + math.sqrt(dis))/(2*a)
        x2 = (-b - math.sqrt(dis))/(2*a)
        return x1, x2
    elif dis == 0:
        x0 = -b /(2*a)
        return x0
    else:
        print("Oups, no solution !")

ii) Test the function `QuadEqu` for:
  - $a=1, b=2, c=1$
  - $a=2, b=2, c=40$
  - $a=2, b=3, c=-5$

In [48]:
# Your code here
QuadEqu(2, 3, -5)

(1.0, -2.5)

c) Write a Python program named `fact2nddeg` that repeatedly offers the user a menu with three options:
- Compute a factorial using factorial2
- Solve a quadratic equation using QuadEqu
- Exit the program

The program should:
- Ask the user to select an option.
- Request the necessary input values for the chosen option.
- Call the corresponding function and display the result.
- Repeat the process until the user chooses to exit.

In [ ]:
# Your code here
def fact2nddeg():
    user_choice = input("""What do you want to do (select a number):
        1. Compute a factorial using factorial2
        2. Solve a quadratic equation using QuadEqu
        3. Exit the program
    """)
    if user_choice == 1:
        n = input("Give me a number then: ")
        n = int()
        fact2nddeg()
    while user_choice == 1
        if user_choice == 1:
        n = input("Give me a number then: ")
        n = int(n)
        fact2nddeg(n)

### Exercise 2
This exercise builds the essential mathematical tools used in public-key cryptography, like [**Rivest–Shamir–Adleman (RSA)**](https://en.wikipedia.org/wiki/RSA_cryptosystem). Security relies on the unique properties of prime numbers and the extreme difficulty of factoring very large numbers.

The Core Principle: Prime Factoring is Hard
   - **Easy**: Multiplying two large prime numbers ($p$ and $q$) to get a huge composite number ($N=p \cdot q$) is very fast for computers.
   - **Hard**: Given only the product $N$, finding the original two prime factors, $p$ and $q$, takes an enormous amount of time. The time required grows exponentially with the size of $N$. This difficulty is what guarantees security.

a) Write a function called `number_of_factors` that takes a positive integer $n$ and returns how many factors the number has together with a list of its factors.

In [5]:
# Your code here
def number_of_factors(n):
    prime_factors = [i for i in range(1, n + 1) if n % i == 0]
    return prime_factors

In [6]:
number_of_factors(12)

[1, 2, 3, 4, 6, 12]

b) Write a Python function called `is_prime` that takes a positive integer, $n$, as input and returns the boolean True if $n$ is a prime number and False, otherwise.

Note : A prime number (or a prime) is a natural number greater than 1 and that has no positive divisors other than 1 and itself.

In [7]:
# Your code here
def is_prime(n):
    if n < 2:
        return False
    elif n == 2:
        return True
    else:
        for i in range(2, n):
            if n % i == 0:
                return False
        return True

c) Cryptographic systems need a large pool of primes. Write a function called `primes(k)` that returns a list of the first k prime numbers. (Set the default value of $k$ to 100).

In [20]:
# Your code here
def primes(k):
    k_list = []
    n = 0
    while len(k_list) < k:
        if is_prime(n):
            k_list.append(n)
        n += 1
    return k_list

In [21]:
primes(5)

[2, 3, 5, 7, 11]

d) Modify the `primes` function to accept an optional argument called start (default value is 2). The function should now return the first $k$ prime numbers that are greater than or equal to start. This simulates searching for primes in a specific high-security range.

In [22]:
# Your code here
def primes(k, start=2):
    k_list = []
    n = start
    while len(k_list) < k:
        if is_prime(n):
            k_list.append(n)
        n += 1
    return k_list

In [23]:
primes(10, 6)

[7, 11, 13, 17, 19, 23, 29, 31, 37, 41]

e) The [**Fundamental theorem of arithmetic**](https://en.wikipedia.org/wiki/Fundamental_theorem_of_arithmetic) states that every integer $n$ is a prime number or can be uniquely written as a product of prime powers (e.g., $12=2^2 \cdot 3^1$. We are interested in finding the exponent of a specific prime in such a factorization. Write a function `V(n, p)` that takes a positive integer $n$ and a prime $p$, and returns the largest integer $r$ such that $p^r$ divides $n$. This value, $r$, is known as the [**p-adic valuation**](https://en.wikipedia.org/wiki/P-adic_valuation).

In [26]:
# Your code here
def V(n, p):
    r = 0
    while n % (p ** r) == 0:
        r += 1
    return r - 1

V(12, 11)

0

f) We would like now to compare the prime-factor structure of two numbers, $n$ and $m$, that is, we would like to measure how different two numbers are in terms of the powers of the primes in their factorizations. Write a function `D(n, m)` that takes two positive integers, $n$ and $m$, and returns a total score $S_1$ calculated as follows:

 $$ S_1 = \sum_{p\; prime} \Bigl| \nu_p(n) - \nu_p(m)\Bigr| $$

Note that `D(14,24)=4`.



In [36]:
# Your code here
def D(n, m):
    g = max(n, m)
    s = 0
    for i in range(g+1):
        if is_prime(i):
            s += abs(V(n, i) - V(m, i))
    return s

In [37]:
D(14, 24)

4

g) The final problem requires you to analyze the average behavior of the `D(n,m)` function by computing a large, iterated sum.  Hence, write a function `P` that takes a positive integer `N` and returns the sum $S_2$ given by:

$$S_2  = \sum_{n=1}^{N}\sum_{m=1}^{N}D(n,m).$$

Note that $P(10)=210$ and $P(10^2) = 37018$.


In [38]:
# Your code here
def P(N):
    s = 0
    for i in range(1, N+1):
        for j in range(1, N+1):
            s += D(i, j)
    return s

In [39]:
P(10)

210

### Exercise 3

Imagine a population of bacteria that grows at a rate proportional to its current size, but with limited food supply. Suppose its growth is modeled by

$$\dfrac{dy}{dt}=0.5y$$

with initial population $y(0)=2$.

a) Explain in words what this equation means about how the population changes over time.

**Your explanation here**

It means that the population of bacteria grows exponentially over time. Moreover, per time unit, the popuplation increases by half its size.

b) [**Euler’s method**](https://en.wikipedia.org/wiki/Euler_method) is a numerical technique to approximate solutions of ordinary differential equations (ODEs) when an exact solution is hard (or impossible) to find.

Suppose we have an ODE:
$$\dfrac{dy}{dt} = f(t, y), \quad y(t_0)=y_0$$

The idea is:

- Start from an initial point $(t_0, y_0)$
- Move forward in small steps of size $h$
- At each step, estimate the new value of $y$ using the slope $f(t,y)$ at the current point

The update formula is thus given by:
$$y_{n+1} = y_n + h \cdot f(t_n,y_n)$$
$$t_{n+1} = t_n + h$$

The pseudocode is given as follows:

c) i) Write $f(t, y) = 0.5y$ as a lambda function.

In [40]:
f = lambda y: 0.5 * y 

ii) Now, use Euler’s method with a step size $h=0.1$ to approximate the population at $t=0.3$. Strictly use lists.

In [47]:
f_1 = 2 + 0.1 * f(2)
f_2 = f_1 + 0.1 * f(3)
f_3 = f_2 + 0.1 * f(4)
f_3


d) Compare your result at $t=0.3$ with the exact solution $y(t) = 2\exp^{0.5t}$, that is, compute the absolute error.

In [49]:
abs(f_3 - 2 * math.exp(0.5 * 0.3))

0.22366848545656604

e) Compare your answers for $y(0.3)$ with $h=0.1$ and $h=0.05$. Which one is closer to the exact solution $y(0.3) = 2\exp^{0.15}$ ? Reflect on this.

In [51]:
f_1 = 2 + 0.05 * f(2)
f_2 = f_1 + 0.05 * f(3)
f_3 = f_2 + 0.05 * f(4)
f_3

2.225

### Exercise 4

Determine how many Sundays fell on the first day of a month during the twentieth century, that is, from 1 January 1901 to 31 December 2000.

The following information might be useful (though you may also verify or research independently):

- 1 January 1900 was a Monday.
- The number of days in each month is as follows:
    - September, April, June, and November each have 30 days.
    - All other months have 31 days, except February, which has 28 days in a common year and 29 days in a leap year.
- Leap year rules:
    - A year is a leap year if it is divisible by 4.
    - However, if the year is a century (i.e., divisible by 100), it is only a leap year if it is also divisible by 400.

In [ ]:
# Your code here
